### DISCUSSION TITLE
Nested absolute value notation?

### DISCUSSION THREAD

**Discussion by motiwalam** [tree-sitter/tree-sitter#1762] [Q&A]
I'm writing a grammar for personal use, and want to use the `|x|` notation for absolute value. 

The rule I have is: `seq('|', $.expr, '|')`, but it does not nest correctly. 

Ideally, the input `||x||` should produce the tree:
```
(abs
    (abs
        (sym)))
```

but it errors. 

The `expr` rule does contain the `abs` rule, so I do not believe that's the problem.

I presume this is because tree-sitter does not know how to distinguish a `|` from one that opens the absolute value rule and one that closes it. 

Is it possible to achieve this behaviour in tree-sitter?

I know that tree-sitter only professes to be capable of parsing CFGs, so I guess an even more general version of my question might be: is this behaviour possible in a context free grammar?

P.S I know this is probably not the best place to ask a question like this. Is there a dedicated forum where I can ask for help?

**Comment by ahelwer** [comment]
Do you have a `||` token in your language? What is the parser debug output?

**Reply by motiwalam** [reply]
I do not. It's just a grammar for arithmetic. The `|` character itself does not appear anywhere except in this rule for the absolute value.

```
$ tree-sitter parse -d <(echo '||x||')
new_parse
process version:0, version_count:1, state:1, row:0, col:0
lex_internal state:1, row:0, column:0
  consume character:'|'
lexed_lookahead sym:|, size:1
shift state:3
process version:0, version_count:1, state:3, row:0, col:1
lex_internal state:1, row:0, column:1
  consume character:'|'
lexed_lookahead sym:|, size:1
shift state:3
process version:0, version_count:1, state:3, row:0, col:2
lex_internal state:1, row:0, column:2
  consume character:'x'
lexed_lookahead sym:symbol, size:1
shift state:42
process version:0, version_count:1, state:42, row:0, col:3
lex_internal state:0, row:0, column:3
  consume character:'|'
lexed_lookahead sym:|, size:1
reduce sym:_expression, child_count:1
shift state:31
process version:0, version_count:1, state:31, row:0, col:4
lex_internal state:0, row:0, column:4
  consume character:'|'
lexed_lookahead sym:|, size:1
reduce sym:absvalue, child_count:3
shift state:3
process version:0, version_count:1, state:3, row:0, col:5
lex_internal state:1, row:0, column:5
  skip character:10
lex_internal state:0, row:0, column:5
  skip character:10
lexed_lookahead sym:end, size:1
detect_error
resume version:0
recover_with_missing symbol:symbol, state:42
recover_to_previous state:29, depth:2
recover_eof
process version:1, version_count:4, state:42, row:0, col:5
reduce sym:_expression, child_count:1
detect_error
process version:2, version_count:4, state:60, row:0, col:5
lex_internal state:0, row:0, column:5
  skip character:10
lexed_lookahead sym:end, size:1
detect_error
process version:3, version_count:4, state:29, row:0, col:5
reduce sym:_expression, child_count:1
detect_error
resume version:0
recover_with_missing symbol:!, state:60
recover_eof
select_smaller_error symbol:ERROR, over_symbol:ERROR
condense
done
(ERROR [0, 0] - [1, 0]
  (absvalue [0, 1] - [0, 4]
    (symbol [0, 2] - [0, 3])))
/dev/fd/63	0 ms	(ERROR [0, 0] - [1, 0])
```

**Comment by motiwalam** [comment]
I have narrowed the problem down. 

The issue was another rule in the grammar allowing for two adjacent expressions to be treated as a multiplication.
In particular, there are two rules:

Rule A, which says a number followed by an absolute value is a multiplication.
Rule B, which says an absolute value followed by another absolute value is a multiplication.

Rule B is responsible for the nesting issue. 
Rule A causes an error in absolute values where the last symbol inside the absolute value is a number (i.e `|x + 1|`).

How can I achieve the semantics of "adjacent terms -> multiplication" without these issues?
As an example, `(x + 1)|x - y|(y + 1)` should get parsed as `(x + 1) * |x - y| * (y + 1)`.
My grammar does in fact parse expressions like the above correctly, but fails in the cases mentioned above.

**Reply by ahelwer** [reply]
Do any of these rules have conflicts or precedences?

**Reply by motiwalam** [reply]
There is one conflict but I don't think it's relevant. Each arithmetic and comparison operation has been given a precedence.

Here's the grammar, https://gist.github.com/thechosenreader/71be282c154f3db93e6b7ed67980400c
